### PR TITLE
Localization Support and CJK Related Optimization

### PR DESCRIPTION
**Pull request by Cyrillya**
# TL;DR
- Port all displayed texts to localized texts handled by `.hjson` files, making it easier to localize.
- CJK (Chinese, Japanese and Korean) character support for `WrapString()`.
- Better [input method](https://en.wikipedia.org/wiki/Input_method) support.

# Localization
Ported all the displayed text in the mod to `.hjson` files to make localization easier.
This pull request also includes full Chinese localization for the mod.

## General
You can get localized texts via vanilla method `Language.GetTextValue()` surely, but for localized text in DragonLens, you can use `LocalizationHelper.GetText()` to get them. You don't have to include the prefix `Mods.DragonLens.` when using this method, shorter and better!

I split localization entries to several `.hjson` files, which can be found in `Localization/` folder.

## Tools
Localization entries for a tool's `DisplayName`, `Description` and `RightClickName` are automatically assigned upon tool registration. The key pattern is `Mods.DragonLens.Tools.{ToolName}.{DataName}`. Here is an example:
```
Mods: {
    DragonLens: {
        Tools: {
            Magnet: {
                DisplayName: Item magnet
                Description: Toggle to pull all items in the world to you. Right click to enable the void magnet, which will automatically destroy items instead.
                RightClickName: Void magnet (destroys items on pickup)
            }
        }
    }
}
```
Actually I made tool localization to a single file named `{LangName}._Mods.DragonLens.Tools.hjson`, so the entries appear like this:
```
> In en-US_Mods.DragonLens.Tools.hjson
Magnet: {
	DisplayName: Item magnet
	Description: Toggle to pull all items in the world to you. Right click to enable the void magnet, which will automatically destroy items instead.
	RightClickName: Void magnet (destroys items on pickup)
}
```
When adding the tool, after you create the class you should compile and reload the mod once, then localization entries will automatically show up in localization files (usually in files under `Localization\Tools\`). No need to write the entries yourself! (though you still have to edit them)

`ToolName` should be the internal name of the tool by default. It can be changed via overriding `LocalizationKey` property. In this example, `ToolName` will be `PlayerEditor` and the key pattern will be `Mods.DragonLens.Tools.PlayerEditor.{DataName}`
```CSharp
public override string LocalizationKey => "PlayerEditor";
```

### Tool Keybinds
Each tool gets its keybind, and some of them have right click keybind as well

The localization key pattern is `Mods.DragonLens.Keybinds.{ToolName}.DisplayName` for left click keybind, and `Mods.DragonLens.Keybinds.{ToolName}RightClick.DisplayName` for right click keybind. Still, entries will be generated in files under `Localization\Keybinds\` when the mod is recompiled and reloaded.

In order for the keybind name to follow the tool name, the value of tool keybinds should be editted only in `en-US` localization file in this pattern:
```
ToolName.DisplayName: "{$Mods.DragonLens.Tools.ToolName.DisplayName}"
ToolNameRightClick.DisplayName: "{$Mods.DragonLens.Tools.ToolName.RightClickName}"
```
You may see commented lines in other languages' localization files. They are auto-generated by tModLoader. Just leave them there.

### Extra keybinds
Only keybinds registered in `ExtraKeybindSystem.cs` should be translated manually.
e.g. in `en-US_Mods.DragonLens.Keybinds.hjson`:
```
CollapseAll.DisplayName: Collapse All
```
Corresponding to `zh-Hans_Mods.DragonLens.Keybinds.hjson`:
```
CollapseAll.DisplayName: 收起全部侧栏
```

## Filters & Separators
Force using localization keys. For filters, `name` and `description` field is gone. Virtual properties, `Name` and `Description`, is in replace with them. They are used for filters automatically get their names and descriptions like `NPCModFilter`, `ItemModFilter`, etc. Override these two properties to adjust text yourself.

`AddSeperator` and `AddFilter` contains `localizationKey` parameter instead of `name` and `description` now. You don't have to include the `Mods.DragonLens.` prefix for localization keys. Here is an example:
```CSharp
filters.AddSeperator("Tools.DustSpawner.FilterCategories.Mod");
filters.AddFilter(new Filter("DragonLens/Assets/Filters/Vanilla", "Tools.DustSpawner.Filters.Vanilla", n => !(n is DustButton && (n as DustButton).dust.type <= DustID.Count)));
```
In the localization file, do this:
```
> en-US_Mods.DragonLens.Tools.hjson
DustSpawner: {
	FilterCategories.Mod: Mod filters
	Filters: {
		Vanilla: {
			Name: Vanilla
			Description: Dusts from the base game
		}
	}
}
```

## Editors
There aren't much changes on them. When declaring them just use `LocalizationHelper.GetText()` to get the names and descriptions, and for some of them we may use `Activator.CreateInstance`. **Do note that** some editors are initialized during mod load and have localized texts. For them we have to update their names and descriptions in methods like `Update()` and `Draw()`. Just as what I did in `DustSpawner.cs`

If they are not updated, then if I change the game language after the mod is loaded, their text will not update.

# CJK Related Optimization
Optimization on input method and word-wrapping

## Word-Wrapping
Since sentences in Chinese, Japanese and Korean don't include spaces, the old word-wrapping method fails to wrap the sentence. They should not be this long:
image
I changed how the method works, mainly the way it splits words.

Each character in CJK counts as a "word" now. If a character has a punctuation after it, the character with the punctuation will be considered a "word", therefore no punctuation appears at the beginning of a line. However, if it is a right-opening punctuation (e.g. left bracket `（`), it will not be connected with the previous character, making it possible to appear at the beginning of the line.

The way it split English (and non-CJK languages) words is slightly changed. The space (if exists) before the word will be included as parts of the "word", since spaces are no longer added to the text when words are combined together to form the final text.

For more detailed changes, see comments in `GUIHelper.WrapString()`

### Displaying
In order to type CJK characters, people use input methods (IME). IME shows up like this in the vanilla game:
image
Two parts of it is missing in DragonLens input bars:
- The IME panel, shows candidate characters.
- The yellow text, that you typed to "search for" the character you want. It is called `CompositionString` in vanilla code.

And I added them back
image

### Bug Fix
Typing with composition string has a bug in vanilla. When you hit backspace, it will delete a character both in the text and the composition string instead of only the one in composition string:
dotnet_d8QkUuAmcy
I made a fix to this for input bars in DragonLens by force-keeping the typed text when composition string exists. See comments in `TextField.cs` for a detailed information.

### CODE REVIEW

File: Content/GUI/Browser.cs [hunk 1]
```diff
@@ -318,16 +321,33 @@ public override void Draw(SpriteBatch spriteBatch)
 			{
 				GUIHelper.DrawOutline(spriteBatch, GetDimensions().ToRectangle(), ModContent.GetInstance<GUIConfig>().buttonColor.InvertColor());
 				HandleText();
+
+				// draw ime panel, note that if there's no composition string then it won't draw anything
+				Main.instance.DrawWindowsIMEPanel(GetDimensions().Position());
```
Comment (ScalarVector1) [inline Content/GUI/Browser.cs:326]: Potential future concern: do we perhaps want to have our own IME panel that respects theme options in the mod config? (color, box style, etc)

File: Content/Tools/Editors/ItemEditor.cs [hunk 2]
```diff
@@ -136,27 +133,29 @@ public void SetupNewItem()
...
-			basicEditorList.Add(new IntEditor("Use Time", n => item.useTime = n, item.useTime, () => item.useTime, "How many ticks between item uses. Ignores input."));
-			basicEditorList.Add(new IntEditor("Use Animation", n => item.useAnimation = n, item.useAnimation, () => item.useAnimation, "How many ticks before you can cancel this items use with input/use it again."));
-			basicEditorList.Add(new BoolEditor("Auto Reuse", n => item.autoReuse = n, item.autoReuse, () => item.autoReuse, "If this item is automatically re-used while holding LMB."));
-			basicEditorList.Add(new IntEditor("Reuse Delay", n => item.reuseDelay = n, item.reuseDelay, () => item.reuseDelay, "The delay between when this item finishes being used and can be used again."));
+			basicEditorList.Add(new IntEditor(GetLocalizedText("UseStyle.Name"), n => item.useStyle = n, item.useStyle, () => item.useStyle, GetLocalizedText("UseStyle.Description")));
```
Comment (ScalarVector1) [inline Content/Tools/Editors/ItemEditor.cs:142]: Thought: Since these all use keys of the form X.name / X.description, could we perhaps collapse the two parameters here to only take the prefix of the key and fetch .Name and .Description automatically?

File: Content/Tools/Gameplay/FastForward.cs [hunk 3]
```diff
@@ -12,19 +12,13 @@ internal class FastForward : Tool
...
 		public override bool HasRightClick => true;
 
-		public override string RightClickName => "Decrease fast forward rate";
-
 		public override void OnActivate()
 		{
 			if (Main.netMode != NetmodeID.SinglePlayer)
 			{
-				Main.NewText("Fast forward is disabled in multiplayer", Color.Red);
+				Main.NewText(LocalizationHelper.GetToolText("Difficulty.MultiplayerDisabled"), Color.Red);
```
Comment (ScalarVector1) [inline Content/Tools/Gameplay/FastForward.cs:21]: Is it correct to re-use the key from the difficulty tool here? Potential oversight

If we want to minimize keys perhaps just a 'This is disabled in multiplayer' key used across all tools would probably be fine

File: Content/Tools/Spawners/BuffSpawner.cs [hunk 4]
```diff
@@ -195,6 +197,10 @@ public override void Draw(SpriteBatch spriteBatch)
 				spriteBatch.Draw(tex, Main.MouseScreen + Vector2.One * 8, new Rectangle(0, 0, tex.Width, tex.Height), Color.White * alpha, 0, default, 1, 0, 0);
 			}
 
+			// Set name here to receive game language selection changes in real time
+			// This is a bit of a hack, but it works
+			durationEditor.name = BuffSpawner.GetText("DurationEditor");
```
Comment (ScalarVector1) [inline Content/Tools/Spawners/BuffSpawner.cs:202]: I wonder if perhaps a more sensible solution is to have a watchdog-type system that will re-initialize all of the UI when the language updates? This is likely an infrequent event so would probably not be a performance issue, and would allow the behavior to be centralized and automatic so there is no pitfall of forgetting to add this when implementing a new tool for example.

File: Helpers/GUIHelper.cs [hunk 5]
```diff
@@ -90,20 +92,71 @@ public static Color InvertColor(this Color color)
 		/// <returns>Input text with linebreaks inserted so it obeys the width constraint.</returns>
 		public static string WrapString(string input, int length, DynamicSpriteFont font, float scale)
 		{
+			// If the character is a CJK character (Chinese, Korean and Japanese), which should be treated as a single word
+			static bool IsCjkPunctuation(char a) => Regex.IsMatch(a.ToString(), @"\p{IsCJKSymbolsandPunctuation}|\p{IsHalfwidthandFullwidthForms}");
```
[the remark this answers is not in the document]
Comment (ScalarVector1) [reply Helpers/GUIHelper.cs:96]: Additionally we probably shouldnt be using arrow notation for methods -- though running the styling cleanup tool will likely clean this up automatically

**Review by ScalarVector1 — APPROVED** [review]
This looks great overall, thank you so much for doing this! I do have a few comments/concerns that may be worth addressing before merging this in, but overall it is probably OK to go in after a quick run of the styling formatter.

File: Content/Tools/Editors/ItemEditor.cs (hunk 2 above)
Comment (Cyrillya) [reply Content/Tools/Editors/ItemEditor.cs:142]: Some editors are auto-generated via `Activator.CreateInstance`, since I don't know much about this approach I can't sure if it works

File: Content/Tools/Spawners/BuffSpawner.cs (hunk 4 above)
Comment (steviegt6) [reply Content/Tools/Spawners/BuffSpawner.cs:207]: We can listen to an event in I believe `LanguageManager` to know when the language changes; flagging for initialization is trivial there.

File: Content/GUI/Browser.cs (hunk 1 above)
Comment (steviegt6) [reply Content/GUI/Browser.cs:326]: More specifically, it might be better to detour the drawing logic and apply our theme changes accordingly when the context is within a DragonLens UI. Depends on how easy that ends up being.

File: Content/Tools/Spawners/BuffSpawner.cs (hunk 4 above)
Comment (ScalarVector1) [reply Content/Tools/Spawners/BuffSpawner.cs:207]: Sounds like a plan. Lets create an issue to look into this

**Comment by Cyrillya** [comment]
I've done my changes. It is now ready for merging or running the "styling formatter"
